### PR TITLE
Discord integration

### PR DESCRIPTION
**Pull request by Devorein**
NOTE: Please dont forget to run `npx prisma migrate dev` to run the latest migrations. I've added a `discord` property to the `User` model that is of `Json` type. We decided to use a json value as we are still not sure which fields to keep store on our database

1. Click on the account modal on the top right
2. Click on discord `Connect` button
3. After successful Login you'll be redirected to previous url
4. Check the account modal (If you don't have an ENS name the discord username will be shown)
5. Since we don't support custom avatar yet its taken from discord if you have one otherwise the first letter of your address

Places where discord account information is being used
1. Bounty reviewer list
2. Bounty assignee and reviewer card
3. Bounty applicants list
4. `@` inside CharmEditor
5. Account card modal
6.  Current Contributors under `/settings/contributors`
7. Focalboard person property

### CODE REVIEW

File: lib/users/getDisplayName.ts [hunk 1]
```diff
@@ -1,9 +1,11 @@
...
+import { DiscordUser, LoggedInUser, User } from 'models/User';
 import { shortenHex } from 'lib/utilities/strings';
 
-export function getDisplayName (user?: User | LoggedInUser) {
+export function getDisplayName (user?: User | LoggedInUser, showDiscriminator?: boolean) {
+  const discordData = user?.discord as unknown as DiscordUser;
+  showDiscriminator = showDiscriminator ?? false;
   if (!user) return '';
-  return (<LoggedInUser> user).ensName || shortenHex(user.addresses[0]);
+  return (<LoggedInUser> user).ensName || discordData ? `${discordData.username}${showDiscriminator ? `#${discordData.discriminator}` : ''}` : shortenHex(user.addresses[0]);
```
[the remark this answers is not in the document]
Comment (mattcasey) [reply lib/users/getDisplayName.ts:8]: I don't see us using showDiscriminator=true anywhere. Let's just not include discriminator in the username and get rid of the option. make our lives easier.. it's in the JSON if we decide we need to use it one day

**Review by mattcasey — CHANGES_REQUESTED** [review]

**Comment by mattcasey** [comment]
My requests for the data
- add a new prisma table "DiscordUser" with 'userId' and 'account' columns, userId points to the user table and 'account' is a JSON blob we get from discord.  This is *not* to be read from except on an account settings page. Keeps our payload sizes down and we don't liter our codebase with the data structure from Discord
- add 'username' and 'avatar' to the user profile. these get updated when you connect discord. In the future we'll let you set 'avatar' by uploading an image to S3. This keeps the logic simple everywhere we look at users, since a user should only have one avatar and one username anyway.
- i think you could pull in the 'discordUser' when we call /api/profile, that way you can tell if they've connected discord or not.

File: lib/users/getDisplayName.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Devorein) [reply lib/users/getDisplayName.ts:8]: Got it. I will make the necessary changes asap

File: adapters/http/index.ts [hunk 2]
```diff
@@ -46,13 +46,13 @@ export function DELETE<T> (
 
 export function POST<T> (
   requestURL: string,
-  data: Params = {},
-  { headers = {}, noHeaders }: { headers?: any, noHeaders?: boolean } = {}
+  data: Params | string = {},
+  { headers = {}, noHeaders, skipStringifying }: { headers?: any, noHeaders?: boolean, skipStringifying?: boolean } = {}
```
[the remark this answers is not in the document]
Comment (Devorein) [reply adapters/http/index.ts:50]: Discord's api accepts `application/x-www-form-urlencoded` post body, so I had to use the `URLSearchParams` to construct the body as per [this](https://stackoverflow.com/questions/67365624/discord-ouath2-access-token-grant-type-none-is-not-supported#fromHistory) StackOverflow thread. Also we had to add the skipStringifying option for the same reason

File: components/common/page-layout/Account/components/AccountModal/AccountModal.tsx [hunk 3]
```diff
@@ -35,25 +52,67 @@ function AccountModal ({ isOpen, onClose }: { isOpen: boolean, onClose: () => vo
...
+      >
+        <Typography color='secondary'>
+          {connectedWithDiscord ? 'Connected with Discord' : 'Connect with Discord'}
+        </Typography>
+        <StyledButton
+          size='small'
+          variant={connectedWithDiscord ? 'contained' : 'outlined'}
+          color={connectedWithDiscord ? 'error' : 'primary'}
+          disabled={isDisconnecting}
+          onClick={async () => {
```
[the remark this answers is not in the document]
Comment (Devorein) [reply components/common/page-layout/Account/components/AccountModal/AccountModal.tsx:94]: Got it

File: prisma/schema.prisma [hunk 4]
```diff
@@ -106,6 +106,18 @@ model User {
   inviteLinks     InviteLink[]
   bounties        Bounty[]
   applications    Application[]
+  username        String?
```
Comment (mattcasey) [inline prisma/schema.prisma:109]: perfect!

**Review by mattcasey — APPROVED** [review]

**Pull request merged**